src/debug.c

```diff
@@ -1819,23 +1821,115 @@ void closeDirectLogFiledes(int fd) {
 }
 
 #ifdef HAVE_BACKTRACE
+#define BACKTRACE_MAX_SIZE 100
+#ifdef __linux__
+#include <sys/prctl.h>
+#include <sys/syscall.h>
+#include <dirent.h>
+
+static pid_t *get_ready_to_signal_threads_tids(pid_t pid, int sig_num, size_t *tids_len_output);
+
+#define MAX_BUFF_LENGTH 256
+typedef struct {
+    char thread_name[16];
+    int trace_size;
+    pid_t tid;
+    void *trace[BACKTRACE_MAX_SIZE];
+} stacktrace_data;
+
+static void *collect_stacktrace_data(void) {
+    /* allocate stacktrace_data struct */
+    stacktrace_data *trace_data = zmalloc(sizeof(stacktrace_data));
+
+    /* Get the stack trace first! */
+    trace_data->trace_size = backtrace(trace_data->trace, BACKTRACE_MAX_SIZE);
+    
+    /* get the thread name */
+    prctl(PR_GET_NAME, trace_data->thread_name);
+
+    /* get the thread id */
+    trace_data->tid = syscall(SYS_gettid);
+
+    /* return the trace data */
+    return trace_data;
+}
+
+static void writeStacktraces(int fd, int uplevel) {
+    /* get the list of all the process's threads that don't block or ignore the THREADS_SIGNAL */
+    pid_t pid = getpid();
+    size_t len_tids = 0;
+    pid_t *tids = get_ready_to_signal_threads_tids(pid, THREADS_SIGNAL, &len_tids);
+
+    /* This call returns either NULL or the stacktraces data from all tids */
+    stacktrace_data **stacktraces_data = (stacktrace_data **)ThreadsManager_runOnThreads(tids, len_tids, collect_stacktrace_data);
+
+    /* free tids */
+    zfree(tids);
+
+    /* ThreadsManager_runOnThreads returns NULL if it is already running */
+    if (!stacktraces_data) return;
+
+
+    char buff[MAX_BUFF_LENGTH];
+    pid_t calling_tid = syscall(SYS_gettid);
+    /* for backtrace_data in backtraces_data: */
+    for (size_t i = 0; i < len_tids; i++) {
+        stacktrace_data *curr_stacktrace_data = stacktraces_data[i];
+        /*ThreadsManager_runOnThreads might fail to collect the thread's data */
+        if (!curr_stacktrace_data) continue;
+        
+        /* stacktrace header includes the tid and the thread's name */
+        snprintf(buff, MAX_BUFF_LENGTH, "\n%d %s", curr_stacktrace_data->tid, curr_stacktrace_data->thread_name);
```

Comment:
Also `snprintf()` is not thread safe, which is why `serverLogFromHandler()` avoids using it.